### ISSUE TITLE
Definition for the support of a function

### ISSUE THREAD

**Issue by avekens** [metamath/set.mm#809]
While working for Linear Algebra, I often had/have to use functions with finite support. Currently, this is expressed like ( `' F " ( _V \ { Z } ) ). Although I used it a lot of times, I am still always puzzled by this cryptic expression. Therefore, I would like to propose a special definition for the support of  functions, as I created already in my Mathbox:

df-supp $a |- supp = ( x e. _V , z e. _V |-> { i e. dom x | ( x ` i ) =/= z } ) $.

By this, ( F supp Z ) can be used instead of ( `' F " ( _V \ { Z } ) ) (under certain, but usually given conditions), which is shorter and more concise. The equivalence of both terms is shown in ~ suppimacnv.

If we can agree on this definition, I would suggest to place it in section "2.3.12  Functions", and to replace all occurences of terms like ( `' F " ( _V \ { Z } ) )  by ( F supp Z ) in the following theorems of (the main part of) set.mm. If I counted correctly, there are about 80 such terms to be replaced.

Would this be OK?

**Comment by benjub** [comment]
Maybe you could replace ( x ` i ) =/= z with ( x " i ) =/= { z } and then prove ~ suppimacnv with the hypothesis relaxed from ` Fun F ` to ` Rel F ` ?

In the comment of df-supp, there is an extra closing parenthesis in the link to the Wikipedia article.

**Comment by david-a-wheeler** [comment]
I like the idea of replacing a cryptic sequence with a clear definition. So I like this one.

**Comment by avekens** [comment]
Good idea, Benoit! Although it was not easy, I could prove ~ suppimacnv using your version of the definition, even without additional hypotheses (except assuring sethood):
      df-supp $a |- supp = ( x e. _V , z e. _V |-> { i e. dom x | ( x " { i } ) =/= { z } } ) $.
      suppimacnv $p |- ( ( R e. V /\ Z e. W ) -> ( R supp Z ) = ( `' R " ( _V \ { Z } ) ) ) $=
I'll update my mathbox accordingly soon.

**Comment by benjub** [comment]
@avekens  Nice !  There's a typo I'll let you fix in a future PR, since it is in your mathbox: to fix the link in the comment, you need to add a space:
~ https://en.wikipedia.org/wiki/Support_(mathematics))
to
~ https://en.wikipedia.org/wiki/Support_(mathematics) )

**Comment by avekens** [comment]
I would like to start with moving the definition of the support of a function and the basic theorems to the main part, adapting the theorems containing the support of a function afterwards now. Benoit and David already agreed to this. @digama0 , @nmegill what is your opinion?

By the way, it is often claimed/required that the support is finite: ( R supp Z ) e. Fin .
Would it be a good idea to add a special definition for this, e.g.

df-finsupp $a |- finSupp = { <. r , z >. | ( r supp z ) e. Fin } $.

By this, `( R supp Z ) e. Fin ` could be written as ` R finSupp Z `, which is (a little bit) shorter and more concise. What do you think?

**Comment by digama0** [comment]
I would suggest putting some stronger closure conditions on it if you want this to be a set. At the very least `R finSupp Z` should assert that `R` is a function. You could put the domain and codomain in there too, but then it's a set-returning ternary function, and the notation on that is not great.

**Comment by avekens** [comment]
@digama0 good idea, so I will use the definition 
df-finsupp $a |- finSupp = { <. r , z >. | ( Fun r /\ ( r supp z ) e. Fin ) } $.

**Comment by avekens** [comment]
I moved the definition of and the theorems for the support of functions from AV's mathbox to main part now, see PR #881.

For this, I created two new sections in the main part of set.mm:

- "The support of functions" (after  "First and second members of an ordered pair"! Since basic properies ( see ~ suppval ) are based on the Axiom of Union ( usage of ~ dmexg ), these definition and theorems cannot be provided earlier)
- "Finitely supported functions" (after "Finite sets")

The theorems which are based on the current representation ` ( ``' R " ( _V \ { Z } ) ) ` of the support of a function and which are provided in previous sections could be moved into this section to have all related theorems in one section, although they do not depend on the Axiom of Union.  This would be possible because they are not used before.  These theorems are listesd in the header comment of "The support of functions".

Before I proceed, the following points should be clarified/decided:

- Should the theorems about the support of functions contained in previous sections be moved into the new section "The support of functions"?
- Should the usage of these theorems about the support of functions contained in previous sections be replaced by the corresponding NEW theorems based on ' supp '?
- Can these theorems be marked with "New usage is discouraged.", so that only the corresponding NEW theorems will be used in the future?

**Comment by avekens** [comment]
Most of the theorems in sections "Cantor normal form" and "Group sum operation" deal with or are using (finite) supports of functions.  For the group sum, I provided alternaives already in section "Group sum operation (alternative)" in my Mathbox. With PR #982, I provided alternatives for the definition of CNF and related theorems using supp and finSupp, see section "Cantor normal form (alternative)", also in my Mathbox.

The next step would be to exchange the current theorems (in the main part of set.mm) by the theorems in my Mathbox, and to mark the current theorems as "obsolete" (adding OLD to the label and "(New usage is discouraged.)" to the comments, etc. ).

Since this is a huge action (about 90 theorems become obsolete, the theorems using them not counted yet), I want to ask if this is OK for everybody, esp. for @digama0 (you contributed most of the affected theorems).

**Comment by digama0** [comment]
The use of OLD and ALT obscures what the actual changes are. Is the only difference in CNF the use of `finSupp` in place of the previous idiom ``( `' F '' ( _V \ 1o ) ) e. Fin``? If so, it doesn't need to be marked as an ALT, that's just installing the new definition everywhere it is used. Same with `mapfien`; it is harder to see if the new proof is smaller since the old proof is still there. But it doesn't appear to have been completely eliminated - there are non-OLD theorems referencing mapfienOLD so it seems like the work isn't done yet.

The summary of my position is that if the only thing you are doing is replacing ``( `' F '' ( _V \ { A } ) ) e. Fin`` with `F finSupp A`, and you are doing this throughout the entire library, then I'm fine with whatever you need to do to make that happen. If you are doing something else on top of that (like breaking up proofs or renaming things) then that should be made explicit. I'm fine with breaking up `mapfien`, especially if it now follows from some theory of finSupp functions.

**Comment by avekens** [comment]
@digama0 Sorry, I was not very consistent using the naming conventions, because these theorems are still in my Mathbox, and work is not done yet at all. There will be severel steps required to complete the replacement. Therefore, the OLD version will be there for a while...

In the next step, as I said, I will replace the current theorems about CNF (and similar about group sum) by the corresponding ALT theorems of my Mathbox, marking the current versions as OLD, e.g.:
- cantnf (current version) -> cantnfOLD
- cantnfALT -> cantnf (new version)
Both versions will be placed next to each other, so that they can be compared more easily.

And yes, I primarily replace the previous idiom ( `' F '' ( _V \ 1o ) ) e. Fin using equivalent expressions using finSupp. As you already observed, however, in certain cases I break up the proofs into lemmata or extract common facts into separate theorems. If I rename an existing theorem (which could be adequate in rare cases), this will be mentioned in the history at the top of set.mm.

Finally, I must admit that the proofs will not become smaller automatically - I do most of the transformation mainly mechanically, and this is already a lot of work. Only if I find an opportunity to improve the proof (accidentally or there is no easy transformation, so I have to think about an alternative anyway), I do it. So there would be possibilities for improvements even after I finished the transformations.

In summary, I think I can confirm that my endeavors go in the direction you expect.

**Comment by avekens** [comment]
Status after PR #988 :
- Revision of section "Cantor normal form" completed: Obsolete theorems (wit suffix OLD) could be removed safely (without any impact on other theorems) now.
- Some theorems in Mathboxes are still affected:

Mathbox SO (@sorear shall/can I perfom these changes?):
- delete mapfien2OLD (mapfien2 to be used instead)
- delete fsuppeq (frnsuppeq to be used instead)
- pwfi2f1o: use supp and finSupp in hypotheses
- pwfi2en: use finSupp in hypothesis
- frlmpwfi: use mapfien2 instead of mapfien2OLD in proof

Mathbox TA:
- "mapfienOLD" is used by "fcobijfs". (@tirix shall/can I revise this?):
- Section "Countable Sets" contains many terms which could be replaced
  by supp and finSupp!  (@tirix will you revise this sometime?):

(New) open points:
- In sections "Abstract multivariate polynomials" and "Euler's partition theorem" (TA's mathbox), the idioms ` ( ``' F " NN ) ` and ` ( ``' F " NN ) e. Fin ` are used, which could be replaced by ` ( F supp 0 ) ` resp. ` F finSupp 0 ` . @digama0 @tirix : should these idioms be replaced, too? 

Next steps:
-  I will replace the deprecated suppssrOLD,  suppssOLD, suppss2OLD, nn0supp by the corresponding versions using supp and finSupp in section "Abstract multivariate polynomials".
- Afterwards, I plan to revise section "Group sum operation" now (replacing the current theorem by their corresponding alternatives of section "Group sum operation" and marking theprevious versions as obsolete)

**Comment by avekens** [comment]
By PR #1030, the revision of the main part of set.mm concerning supp/finSupp is done:
- obsolete theorems are only used by obsolete theorems
- the sequence " ( _V \ {  (indication the previous definition of the support) is only used in obsolete theorems and the theorems ~ cnvimadfsn , ~ suppimacnvss , ~ suppimacnv , ~ fsuppcolem od section "The support of functions".

The sequence " ( _V \ { (i.e. the previous definition of the support) is still used in some Mathboxes (of TA and SO) who should replace/remove it on occation.

Finally, the obsolete theorems are still used in the Mathboxes of TA and AV. As for my Mathbox, I will replace them in the next days. The remaining usages in TA's Mathbox are:
* ~ fnniniseg2OLD (to be replaced by ~ suppvalfn ): ~ fpwrelmapffslem , ~ fsumcvg4
* ~ fnsuppresOLD : ~ resf1o
* ~ suppssOLD : ~ resf1o
* ~ suppssrOLD : ~ eulerpartlemb
* ~ suppss2OLD : ~ suppss3
* ~ gsumclOLD : ~ gsumle , ~ gsumvsca1 , ~ gsumvsca2 , ~ xrge0tsmsd , ~ gsumesum , ~ esumlub , ~ sitgclg
* ~ gsumf1oOLD : ~ gsummptf1o
* ~ gsumsplitOLD : ~ gsumle , ~ gsummptun , ~ gsumesum
* ~ gsumsplit2OLD : ~ gsumunsnf
* ~ gsummhmOLD : ~ gsummptmhm

@tirix will you replace them in the near future?

After the revision of the Mathboxes is finished, the obsolete theorems could be removed completely.

**Comment by tirix** [comment]
Hi Alexander,
Yes, I would like to replace them, probably at my own pace (which may be slow..). I hope it will be possible to keep the old versions until I do so, and avoid breaking those theorems in my math box.

**Comment by tirix** [comment]
@avekens I think this action is complete and can be closed, except if there are more theorems which still use to the obsolete notation.

**Comment by avekens** [comment]
@tirix what about the 10 theorems in your Mathbox still using the old, obsolete notation? These theorems are:
~ gsumle
~ gsumvsca1
~ gsumvsca2 
~ xrge0tsmsd
~ gsumesum 
~ esumlub 
~ sitgclg
~ gsummptf1o
~ gsummptun
~ gsumunsnf

If you agree, I would like to replace the OLD theorems in the proofs of these theorems, and afterwards I will close this issue.

**Comment by avekens** [comment]
With PR ##1316 , I replaced the OLD theorems in the remaining theorems in @tirix`s Mathbox (see comment on 21 Oct), moving some of them to the main part of set.mm.

By this, the OLD theorems are either used by other OLD theorems only or not used at all now. Therefore, they could be removed on occasion (in May 2020...). 

This issue should be kept open until the removal is done.

**Comment by avekens** [comment]
By PR #1800, the final cleanup is done.
